### PR TITLE
Cache the MarkdownParser & Processors

### PR DESCRIPTION
**Pull request by MihaZupan**
- `MarkdownParser` is made static (had private ctor before)
- `Block/InlineProcessor` got a set of Setup/Rent/Reset/Release methods

```c#
Markdown.Parse("Hello world!");
```

Before:
image

After:
image

### CODE REVIEW

**Comment by MihaZupan** [comment]
I think this makes of free of "Constant overhead" in terms of allocations on the `Parse`/`ToHtml` paths.

File: src/Markdig/Parsers/BlockProcessor.cs [hunk 1]
```diff
@@ -16,165 +17,159 @@ namespace Markdig.Parsers
...
-            Open(document);
-        }
+        public bool TrackTrivia { get; private set; }
 
         public bool SkipFirstUnwindSpace { get; set; }
 
+        private bool ContinueProcessingLine { get; set; }
+
         /// <summary>
-        /// Gets the new blocks to push. A <see cref="BlockParser"/> is required to push new blocks that it creates to this property.
```
Comment (xoofx) [inline src/Markdig/Parsers/BlockProcessor.cs:66]: Why are the order of the properties are all changed in this file? Could this be done in a separate PR?

Comment (MihaZupan) [reply src/Markdig/Parsers/BlockProcessor.cs:66]: Oh yeah I can try to revert this part. I was doing it to triple-check all state was getting reset

Comment (MihaZupan) [reply src/Markdig/Parsers/BlockProcessor.cs:66]: Changed back to original order
